src/libstd/io/mod.rs

```diff
@@ -1661,6 +1661,45 @@ fn read_until<R: BufRead + ?Sized>(r: &mut R, delim: u8, buf: &mut Vec<u8>) -> R
     }
 }
 
+fn read_while<R, P>(r: &mut R, buf: &mut Vec<u8>, mut predicate: P) -> Result<usize>
+where
+    R: BufRead + ?Sized,
+    P: FnMut(u8) -> bool,
+{
+    let mut read = 0;
+    let mut inner_read = 0;
+
+    'outer: loop {
+        let available = match r.fill_buf() {
+            Ok(n) => n,
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => continue,
+            Err(e) => {
+                r.consume(read);
+                return Err(e);
+            }
+        };
+        if available.len() == 0 {
+            break;
+        }
+
+        for byte in available.bytes() {
+            let byte = byte?;
+            if predicate(byte) {
+                buf.extend_from_slice(&[byte]);
```

Comment:
We probably want to search for the cutoff and then do a single extend rather than appending one byte at a time:

```rust
fn read_while<R, F>(r: &mut R, mut predicate: F, out: &mut Vec<u8>) -> io::Result<usize>
where
    R: BufRead,
    F: FnMut(u8) -> bool,
{
    let mut len = 0;
    loop {
        let available = r.fill_buf()?;

        if available.is_empty() {
            return Ok(len);
        }

        let i = available.iter().position(|b| !predicate(*b));

        let cutoff = i.unwrap_or(available.len());
        len += cutoff;
        out.extend_from_slice(&available[..cutoff]);
        r.consume(cutoff);

        if i.is_some() {
            return Ok(len);
        }
    }
}
```

Reply:
Oh yeah that's a lot more elegant; both shorter, more efficient, and gets rid of the labeled breaks. Thanks for writing this out!